tap-agent/src/tap/sender_allocation.rs

```diff
@@ -0,0 +1,587 @@
+// Copyright 2023-, GraphOps and Semiotic Labs.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{str::FromStr, sync::Arc, time::Duration};
+
+use alloy_primitives::Address;
+use alloy_sol_types::Eip712Domain;
+use anyhow::{anyhow, ensure, Result};
+use eventuals::Eventual;
+use indexer_common::{escrow_accounts::EscrowAccounts, prelude::SubgraphClient};
+use jsonrpsee::{core::client::ClientT, http_client::HttpClientBuilder, rpc_params};
+use sqlx::{types::BigDecimal, PgPool};
+use tap_aggregator::jsonrpsee_helpers::JsonRpcResponse;
+use tap_core::{
+    eip_712_signed_message::EIP712SignedMessage,
+    receipt_aggregate_voucher::ReceiptAggregateVoucher,
+    tap_manager::RAVRequest,
+    tap_receipt::{ReceiptCheck, ReceivedReceipt},
+};
+use tokio::sync::Mutex;
+use tracing::{error, warn};
+
+use crate::{
+    config::{self},
+    tap::{
+        escrow_adapter::EscrowAdapter, rav_storage_adapter::RAVStorageAdapter,
+        receipt_checks_adapter::ReceiptChecksAdapter,
+        receipt_storage_adapter::ReceiptStorageAdapter, signers_trimmed,
+        unaggregated_receipts::UnaggregatedReceipts,
+    },
+};
+
+type TapManager = tap_core::tap_manager::Manager<
+    EscrowAdapter,
+    ReceiptChecksAdapter,
+    ReceiptStorageAdapter,
+    RAVStorageAdapter,
+>;
+
+/// Manages unaggregated fees and the TAP lifecyle for a specific (allocation, sender) pair.
+pub struct SenderAllocation {
+    pgpool: PgPool,
+    tap_manager: TapManager,
+    allocation_id: Address,
+    sender: Address,
+    sender_aggregator_endpoint: String,
+    unaggregated_fees: Arc<Mutex<UnaggregatedReceipts>>,
```

Comment:
At this point, why do you have an `Arc<Mutex<UnaggreagatedReceipts>>` instead of just replacing the `unaggregated_fees` itself? It seems that all fields are mutated at once.

Reply:
Yes, but then I have to make many of the methods mutable, which makes usage in `SenderAccount` more cumbersome (I'd have to play with `Mutex<SenderAllocation>` there which only displaces the problem).

Reply:
Ok, in this case I think you should use the blocking mutex for this specific use case. You can take a look on [tokio's documentation](https://docs.rs/tokio/1.36.0/tokio/sync/struct.Mutex.html#which-kind-of-mutex-should-you-use).

We don't have any awaiting point that justifies the need of an async mutex.

Reply:
By using a blocking mutext, we simplify a lot the current logic

Reply:
Fix in [3bba8fd](https://github.com/graphprotocol/indexer-rs/pull/113/commits/3bba8fdc8db32f7deb772cdb444ded422f1182c3) and [1135be5](https://github.com/graphprotocol/indexer-rs/pull/113/commits/1135be5c071031f3d4b2b94d441192c3a0e61877)